File: frontend/src/stores/pipeline.ts
import { create } from 'zustand';
import { devtools, persist } from 'zustand/middleware';
import { apiClient } from '@/api/client';
import type { RFQStatus, Priority } from '@/types';

interface RFQItem {
  id: string;
  rfqNumber: string;
  customerName: string;
  customerId: string;
  title: string;
  description?: string;
  dueDate: string;
  receivedDate: string;
  estimatedValue?: number;
  priority: Priority;
  status: RFQStatus;
  account?: {
    name: string;
    id?: string;
  };
  assignee?: {
    id: string;
    name: string;
    avatar?: string;
  };
  tags: string[];
  version: number;
  attachmentCount: number;
  commentCount: number;
  lastActivityAt: string;
}

interface PipelineStats {
  totalRFQs: number;
  activeRFQs: number;
  totalValue: number;
  avgResponseTime: number;
  conversionRate: number;
  overdueCount: number;
}

interface PipelineState {
  // State
  rfqs: RFQItem[];
  stats: PipelineStats;
  isLoading: boolean;
  error: string | null;
  lastFetchedAt: number | null;

  // Actions
  fetchRFQs: () => Promise<void>;
  fetchRFQById: (id: string) => Promise<RFQItem | null>;
  createRFQ: (rfq: Partial<RFQItem>) => Promise<RFQItem>;
  updateRFQ: (id: string, updates: Partial<RFQItem>) => Promise<RFQItem>;
  deleteRFQ: (id: string) => Promise<void>;
  bulkDeleteRFQs: (ids: string[]) => Promise<void>;
  exportRFQs: (ids?: string[]) => Promise<void>;
  setRFQStatus: (id: string, status: RFQStatus) => Promise<void>;
  assignRFQ: (id: string, assigneeId: string) => Promise<void>;
  clearError: () => void;
}

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000/api/v1';

export const usePipelineStore = create<PipelineState>()(
  devtools(
    persist(
      (set, get) => ({
        // Initial state
        rfqs: [],
        stats: {
          totalRFQs: 0,
          activeRFQs: 0,
          totalValue: 0,
          avgResponseTime: 0,
          conversionRate: 0,
          overdueCount: 0,
        },
        isLoading: false,
        error: null,
        lastFetchedAt: null,

        // Fetch all RFQs
        fetchRFQs: async () => {
          const { lastFetchedAt } = get();
          const now = Date.now();

          // Cache for 30 seconds
          if (lastFetchedAt && now - lastFetchedAt < 30000) {
            return;
          }

          set({ isLoading: true, error: null });
          try {
            const data = await apiClient.get<any>('/rfqs');
            
            // Calculate stats
            const rfqs: RFQItem[] = data.items || [];
            const activeStatuses: RFQStatus[] = ['new', 'reviewing', 'quoting', 'submitted'];
            const activeRFQs = rfqs.filter(rfq => activeStatuses.includes(rfq.status));
            const totalValue = rfqs.reduce((sum, rfq) => sum + (rfq.estimatedValue || 0), 0);
            const overdueCount = rfqs.filter(rfq => new Date(rfq.dueDate) < new Date()).length;
            const wonRFQs = rfqs.filter(rfq => rfq.status === 'won').length;
            const totalSubmitted = rfqs.filter(rfq => rfq.status === 'submitted').length;
            
            // Calculate avg response time (mock for now)
            const avgResponseTime = rfqs.reduce((sum, rfq) => {
              const received = new Date(rfq.receivedDate).getTime();
              const lastActivity = new Date(rfq.lastActivityAt).getTime();
              return sum + (lastActivity - received);
            }, 0) / (rfqs.length || 1) / (1000 * 60 * 60); // Convert to hours

            set({
              rfqs,
              stats: {
                totalRFQs: rfqs.length,
                activeRFQs: activeRFQs.length,
                totalValue,
                avgResponseTime: Math.round(avgResponseTime),
                conversionRate: totalSubmitted > 0 ? Math.round((wonRFQs / totalSubmitted) * 100) : 0,
                overdueCount,
              },
              isLoading: false,
              lastFetchedAt: now,
            });
          } catch (error) {
            console.error('Error fetching RFQs:', error);
            set({
              error: error instanceof Error ? error.message : 'Failed to fetch RFQs',
              isLoading: false,
            });
          }
        },

        // Fetch single RFQ
        fetchRFQById: async (id: string) => {
          try {
            const rfq = await apiClient.get<RFQItem>(`/rfqs/${id}`);

            // Update in store
            set(state => ({
              rfqs: state.rfqs.map(r => r.id === id ? rfq : r),
            }));

            return rfq;
          } catch (error) {
            console.error('Error fetching RFQ:', error);
            set({ error: error instanceof Error ? error.message : 'Failed to fetch RFQ' });
            return null;
          }
        },

        // Create RFQ
        createRFQ: async (rfqData: Partial<RFQItem>) => {
          try {
            const newRFQ = await apiClient.post<RFQItem>('/rfqs', rfqData);

            // Add to store
            set(state => ({
              rfqs: [newRFQ, ...state.rfqs],
              stats: {
                ...state.stats,
                totalRFQs: state.stats.totalRFQs + 1,
                activeRFQs: state.stats.activeRFQs + 1,
              },
            }));

            return newRFQ;
          } catch (error) {
            console.error('Error creating RFQ:', error);
            set({ error: error instanceof Error ? error.message : 'Failed to create RFQ' });
            throw error;
          }
        },

        // Update RFQ
        updateRFQ: async (id: string, updates: Partial<RFQItem>) => {
          const currentRFQ = get().rfqs.find(r => r.id === id);
          if (!currentRFQ) {
            throw new Error('RFQ not found');
          }

          try {
            const updatedRFQ = await apiClient.put<RFQItem>(`/rfqs/${id}`, updates, {
              headers: {
                'If-Match': `"${currentRFQ.version}"`, // Optimistic locking
              },
            });

            // Update in store
            set(state => ({
              rfqs: state.rfqs.map(r => r.id === id ? updatedRFQ : r),
            }));

            return updatedRFQ;
          } catch (error: any) {
            console.error('Error updating RFQ:', error);
            if (error.code === '409' || error.message?.includes('409')) {
               set({ error: 'RFQ was modified by another user. Please refresh and try again.' });
            } else {
               set({ error: error.message || 'Failed to update RFQ' });
            }
            throw error;
          }
        },

        // Delete RFQ
        deleteRFQ: async (id: string) => {
          try {
            await apiClient.delete(`/rfqs/${id}`);

            // Remove from store
            set(state => ({
              rfqs: state.rfqs.filter(r => r.id !== id),
              stats: {
                ...state.stats,
                totalRFQs: state.stats.totalRFQs - 1,
              },
            }));
          } catch (error) {
            console.error('Error deleting RFQ:', error);
            set({ error: error instanceof Error ? error.message : 'Failed to delete RFQ' });
            throw error;
          }
        },

        // Bulk delete RFQs
        bulkDeleteRFQs: async (ids: string[]) => {
          try {
            await apiClient.post('/rfqs/bulk-delete', { ids });

            // Remove from store
            set(state => ({
              rfqs: state.rfqs.filter(r => !ids.includes(r.id)),
              stats: {
                ...state.stats,
                totalRFQs: state.stats.totalRFQs - ids.length,
              },
            }));
          } catch (error) {
            console.error('Error bulk deleting RFQs:', error);
            set({ error: error instanceof Error ? error.message : 'Failed to bulk delete RFQs' });
            throw error;
          }
        },

        // Export RFQs
        exportRFQs: async (ids?: string[]) => {
          try {
            const queryParams = ids && ids.length > 0 ? `?ids=${ids.join(',')}` : '';
            const response = await fetch(`${API_BASE_URL}/rfqs/export${queryParams}`, {
              headers: {
                'Authorization': `Bearer ${localStorage.getItem('access_token')}`,
              },
            });

            if (!response.ok) {
              throw new Error(`Failed to export RFQs: ${response.statusText}`);
            }

            // Download the file
            const blob = await response.blob();
            const url = window.URL.createObjectURL(blob);
            const a = document.createElement('a');
            a.href = url;
            a.download = `rfqs_export_${new Date().toISOString().split('T')[0]}.csv`;
            document.body.appendChild(a);
            a.click();
            window.URL.revokeObjectURL(url);
            document.body.removeChild(a);
          } catch (error) {
            console.error('Error exporting RFQs:', error);
            set({ error: error instanceof Error ? error.message : 'Failed to export RFQs' });
            throw error;
          }
        },

        // Set RFQ status
        setRFQStatus: async (id: string, status: RFQStatus) => {
          await get().updateRFQ(id, { status });
        },

        // Assign RFQ
        assignRFQ: async (id: string, assigneeId: string) => {
          // Fetch assignee details (mock for now)
          const assignee = {
            id: assigneeId,
            name: 'User Name', // Would fetch from API
          };
          await get().updateRFQ(id, { assignee });
        },

        // Clear error
        clearError: () => set({ error: null }),
      }),
      {
        name: 'pipeline-storage',
        partialize: (state) => ({
          // Only persist RFQs and lastFetchedAt
          rfqs: state.rfqs,
          lastFetchedAt: state.lastFetchedAt,
        }),
      }
    ),
    { name: 'PipelineStore' }
  )
);

File: sensei-rs/crates/sensei-frontend/src/stores/pipeline.rs
//! Pipeline (RFQ management) reactive store.
//!
//! Mirrors the Zustand [`pipeline.ts`](frontend/src/stores/pipeline.ts) store.

use crate::api::client::{ApiClient, ApiError};
use crate::api::rfq::{RfqApi, RfqDto};
use leptos::prelude::*;

/// Computed pipeline statistics.
#[derive(Debug, Clone)]
pub struct PipelineStats {
    pub total_rfqs: i32,
    pub active_rfqs: i32,
    pub total_value: f64,
    pub avg_response_time: f64,
    pub conversion_rate: f64,
    pub overdue_count: i32,
}

/// Reactive store for the RFQ pipeline.
#[derive(Debug, Clone)]
pub struct PipelineStore {
    pub rfqs: RwSignal<Vec<RfqDto>>,
    pub current_rfq: RwSignal<Option<RfqDto>>,
    pub stats: RwSignal<Option<PipelineStats>>,
    pub is_loading: RwSignal<bool>,
    pub error: RwSignal<Option<String>>,
    pub last_fetched_at: RwSignal<Option<String>>,
}

impl PipelineStore {
    pub fn new() -> Self {
        Self {
            rfqs: RwSignal::new(Vec::new()),
            current_rfq: RwSignal::new(None),
            stats: RwSignal::new(None),
            is_loading: RwSignal::new(false),
            error: RwSignal::new(None),
            last_fetched_at: RwSignal::new(None),
        }
    }

    fn compute_stats(rfqs: &[RfqDto]) -> PipelineStats {
        let total = rfqs.len() as i32;
        let active = rfqs
            .iter()
            .filter(|r| matches!(r.status.as_str(), "open" | "in_progress" | "quoted"))
            .count() as i32;
        let total_value: f64 = rfqs.iter().filter_map(|r| r.estimated_value).sum();
        let overdue = rfqs
            .iter()
            .filter(|r| {
                r.due_date
                    .as_ref()
                    .is_some_and(|d| *d < chrono::Utc::now().format("%Y-%m-%d").to_string())
            })
            .count() as i32;

        PipelineStats {
            total_rfqs: total,
            active_rfqs: active,
            total_value,
            avg_response_time: 0.0,
            conversion_rate: 0.0,
            overdue_count: overdue,
        }
    }

    /// Fetch all RFQs.
    pub async fn fetch_rfqs(&self, client: &ApiClient) {
        self.is_loading.set(true);
        self.error.set(None);
        match RfqApi::list_rfqs(client, None).await {
            Ok(resp) => {
                self.rfqs.set(resp.items.clone());
                self.stats.set(Some(Self::compute_stats(&resp.items)));
            }
            Err(e) => self.error.set(Some(e.to_string())),
        }
        self.is_loading.set(false);
    }

    /// Fetch a single RFQ by ID.
    pub async fn fetch_rfq_by_id(&self, client: &ApiClient, id: &str) {
        self.is_loading.set(true);
        self.error.set(None);
        match RfqApi::get_rfq(client, id).await {
            Ok(data) => {
                self.current_rfq.set(Some(data.clone()));
                // Also update in the list
                self.rfqs.update(|rfqs| {
                    if let Some(pos) = rfqs.iter().position(|x| x.id == id) {
                        rfqs[pos] = data;
                    }
                });
            }
            Err(e) => self.error.set(Some(e.to_string())),
        }
        self.is_loading.set(false);
    }

    /// Fetch RFQ details (additional info).
    pub async fn fetch_rfq_details(&self, client: &ApiClient, id: &str) {
        self.is_loading.set(true);
        self.error.set(None);
        match client
            .get::<serde_json::Value>(&format!("/api/v1/rfqs/{}/details", id))
            .await
        {
            Ok(_) => {
                // Merge details into current RFQ
                if let Some(current) = self.current_rfq.get() {
                    // Extend with details; current approach: replace with enriched data
                    self.current_rfq.set(Some(RfqDto {
                        id: current.id.clone(),
                        ..current
                    }));
                }
            }
            Err(e) => self.error.set(Some(e.to_string())),
        }
        self.is_loading.set(false);
    }

    /// Create a new RFQ.
    pub async fn create_rfq(
        &self,
        client: &ApiClient,
        data: &serde_json::Value,
    ) -> Result<RfqDto, ApiError> {
        let rfq: RfqDto = client.post("/api/v1/rfqs", data).await?;
        self.rfqs.update(|r| r.push(rfq.clone()));
        Ok(rfq)
    }

    /// Update an existing RFQ.
    pub async fn update_rfq(
        &self,
        client: &ApiClient,
        id: &str,
        updates: &serde_json::Value,
    ) -> Result<RfqDto, ApiError> {
        let rfq: RfqDto = client.put(&format!("/api/v1/rfqs/{}", id), updates).await?;
        self.rfqs.update(|r| {
            if let Some(pos) = r.iter().position(|x| x.id == id) {
                r[pos] = rfq.clone();
            }
        });
        Ok(rfq)
    }

    /// Delete an RFQ.
    pub async fn delete_rfq(&self, client: &ApiClient, id: &str) -> Result<(), ApiError> {
        RfqApi::delete_rfq(client, id).await?;
        self.rfqs.update(|r| r.retain(|x| x.id != id));
        Ok(())
    }

    /// Bulk delete RFQs.
    pub async fn bulk_delete_rfqs(
        &self,
        client: &ApiClient,
        ids: &[String],
    ) -> Result<(), ApiError> {
        let payload = serde_json::json!({ "ids": ids });
        client
            .post::<serde_json::Value, _>("/api/v1/rfqs/bulk-delete", &payload)
            .await?;
        self.rfqs.update(|r| r.retain(|x| !ids.contains(&x.id)));
        Ok(())
    }

    /// Export RFQs as a file (PDF/Excel).
    ///
    /// Uses the shared client (same connection pool, bearer token, and 401
    /// refresh pipeline) instead of constructing a fresh `reqwest::Client`.
    pub async fn export_rfqs(
        client: &ApiClient,
        ids: Option<&[String]>,
    ) -> Result<Vec<u8>, ApiError> {
        let path = match ids {
            Some(id_list) => {
                let ids_str: Vec<&str> = id_list.iter().map(|s| s.as_str()).collect();
                format!("/api/v1/rfqs/export?ids={}", ids_str.join(","))
            }
            None => "/api/v1/rfqs/export".to_string(),
        };
        client.get_bytes(&path).await
    }

    /// Set RFQ status.
    pub async fn set_rfq_status(
        &self,
        client: &ApiClient,
        id: &str,
        status: &str,
    ) -> Result<RfqDto, ApiError> {
        let payload = serde_json::json!({ "status": status });
        let rfq: RfqDto = client
            .put(&format!("/api/v1/rfqs/{}/status", id), &payload)
            .await?;
        self.rfqs.update(|r| {
            if let Some(pos) = r.iter().position(|x| x.id == id) {
                r[pos] = rfq.clone();
            }
        });
        Ok(rfq)
    }

    /// Assign an RFQ to a user.
    pub async fn assign_rfq(
        &self,
        client: &ApiClient,
        id: &str,
        assignee_id: &str,
    ) -> Result<RfqDto, ApiError> {
        let rfq = RfqApi::assign_rfq(client, id, assignee_id).await?;
        self.rfqs.update(|r| {
            if let Some(pos) = r.iter().position(|x| x.id == id) {
                r[pos] = rfq.clone();
            }
        });
        Ok(rfq)
    }

    pub fn clear_error(&self) {
        self.error.set(None);
    }
}

impl Default for PipelineStore {
    fn default() -> Self {
        Self::new()
    }
}
